Add support for processing WAN replication messages from 3.x members

mmedenjak opened the pull request:
Adds support on PASSIVE (target) cluster to be able to deserialize and
process WAN replication events from a 3.x ACTIVE (source) cluster.

Testing is done by running the full WAN suite on 3.12 using compatibility framework to start 4.0 clusters with this change.

EE: https://github.com/hazelcast/hazelcast-enterprise/pull/3578

Corresponding PRs for 4.0->3.12 replication:
https://github.com/hazelcast/hazelcast/pull/16972
https://github.com/hazelcast/hazelcast-enterprise/pull/3597

mmedenjak commented:
run-lab-run

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/602/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/602/Hazelcast-pr-builder-602.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientReliableTopicOnClusterRestartTest.shouldContinue_OnClusterRestart_whenDataLoss_LossTolerant_afterInvocationTimeout:167 CountDownLatch failed to complete within 120 seconds, count left: 1
[ERROR]   ClientReliableTopicOnClusterRestartTest.shouldFail_OnClusterRestart_whenDataLoss_notLossTolerant:214->lambda$shouldFail_OnClusterRestart_whenDataLoss_notLossTolerant$1:216
[ERROR]   DataSerializableImplementsVersionedTest.testIdentifiedDataSerializableForVersionedInterface:85->checkInstanceOfVersion:135 Expected com.hazelcast.internal.compatibility.map.CompatibilityWanMapEntryView to implement Versioned, since in.getVersion() is used
[ERROR] Errors: 
[ERROR]   ServiceProviderTest.remoteServiceLoaderTest:39 ? Hazelcast Service with name '...
[ERROR]   CacheConfigConstructorTest.testConstructor:44 ? Guardian java.lang.ClassNotFou...
[ERROR]   ConfigConstructorTest.testConstructor:50 ? NoSuchField Field groupConfig does ...
[ERROR]   DynamicConfigurationAwareConfigConstructorTest.testConstructor:57 ? NoSuchField
[INFO] 
[ERROR] Tests run: 24160, Failures: 3, Errors: 4, Skipped: 996
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/620/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/620/Hazelcast-pr-builder-620.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.19 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/AbstractSerializationService.java:52:15: Unused import - com.hazelcast.internal.serialization.impl.SerializationConstants.CONSTANT_SERIALIZERS_LENGTH. [UnusedImports]
--------------------------
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/623/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/623/Hazelcast-pr-builder-623.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.19 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/AbstractSerializationService.java:52:15: Unused import - com.hazelcast.internal.serialization.impl.SerializationConstants.CONSTANT_SERIALIZERS_LENGTH. [UnusedImports]
--------------------------
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/624/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/624/Hazelcast-pr-builder-624.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.19 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/AbstractSerializationService.java:52:15: Unused import - com.hazelcast.internal.serialization.impl.SerializationConstants.CONSTANT_SERIALIZERS_LENGTH. [UnusedImports]
--------------------------
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/627/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/627/Hazelcast-pr-builder-627.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientMapEventJournalBasicTest>AbstractEventJournalBasicTest.allowReadingWithFutureSeq:511->HazelcastTestSupport.assertOpenEventually:1222->HazelcastTestSupport.assertOpenEventually:1237 CountDownLatch failed to complete within 30 seconds, count left: 1
[ERROR]   DataSerializableImplementsVersionedTest.testIdentifiedDataSerializableForVersionedInterface:85->checkInstanceOfVersion:135 Expected com.hazelcast.internal.compatibility.map.CompatibilityWanMapEntryView to implement Versioned, since in.getVersion() is used
[ERROR] Errors: 
[ERROR]   ServiceProviderTest.remoteServiceLoaderTest:39 ? Hazelcast Service with name '...
[ERROR]   CacheConfigConstructorTest.testConstructor:44 ? Guardian java.lang.ClassNotFou...
[ERROR]   ConfigConstructorTest.testConstructor:50 ? NoSuchField Field groupConfig does ...
[ERROR]   DynamicConfigurationAwareConfigConstructorTest.testConstructor:57 ? NoSuchField
[INFO] 
[ERROR] Tests run: 24160, Failures: 2, Errors: 4, Skipped: 996
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/630/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/630/Hazelcast-pr-builder-630.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   DataSerializableImplementsVersionedTest.testIdentifiedDataSerializableForVersionedInterface:85->checkInstanceOfVersion:135 Expected com.hazelcast.internal.compatibility.map.CompatibilityWanMapEntryView to implement Versioned, since in.getVersion() is used
[ERROR] Errors: 
[ERROR]   CacheConfigConstructorTest.testConstructor:44 ? Guardian java.lang.ClassNotFou...
[INFO] 
[ERROR] Tests run: 24160, Failures: 1, Errors: 1, Skipped: 998
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/635/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/635/Hazelcast-pr-builder-635.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientMapEventJournalBasicTest>AbstractEventJournalBasicTest.allowReadingWithFutureSeq:511->HazelcastTestSupport.assertOpenEventually:1222->HazelcastTestSupport.assertOpenEventually:1237 CountDownLatch failed to complete within 30 seconds, count left: 1
[INFO] 
[ERROR] Tests run: 24160, Failures: 1, Errors: 0, Skipped: 999
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/641/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/641/Hazelcast-pr-builder-641.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   DataSerializableConventionsTest.test_identifiedDataSerializables_areInstancesOfSameClass_whenConstructedFromFactory:266 Factory with ID -31 instantiated an object of class com.hazelcast.config.WanReplicationConfig while expected type was class com.hazelcast.internal.compatibility.wan.CompatibilityWanReplicationEvent
[ERROR]   DataSerializableConventionsTest.test_identifiedDataSerializables_haveUniqueFactoryAndTypeId:191 Factory-Type ID pair {-31, 0} from class com.hazelcast.config.WanReplicationConfig is already registered in another type.
[ERROR]   PacketTest.setPacketType:50 expected same:<COMPATIBILITY_BIND_MESSAGE> was not:<BIND>
[ERROR] Errors: 
[ERROR]   ProbabilisticSplitBrainProtectionFunctionTest.testSplitBrainProtectionAbsent_whenHeartbeatsLate:67->createSplitBrainProtectionFunctionProxy:120 ? ClassNotFound
[ERROR]   ProbabilisticSplitBrainProtectionFunctionTest.testSplitBrainProtectionPresent_whenHeartbeatsOnTime:79->createSplitBrainProtectionFunctionProxy:120 ? ClassNotFound
[ERROR]   RecentlyActiveSplitBrainProtectionFunctionTest.testSplitBrainProtectionAbsent_whenHeartbeatsReceivedBeforeToleratedWindow:70->createSplitBrainProtectionFunctionProxy:78 ? ClassNotFound
[INFO] 
[ERROR] Tests run: 3807, Failures: 3, Errors: 3, Skipped: 14
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/659/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/659/Hazelcast-pr-builder-659.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientCacheEventJournalBasicTest>AbstractEventJournalBasicTest.allowReadingWithFutureSeq:511->HazelcastTestSupport.assertOpenEventually:1222->HazelcastTestSupport.assertOpenEventually:1237 CountDownLatch failed to complete within 30 seconds, count left: 1
[INFO] 
[ERROR] Tests run: 24160, Failures: 1, Errors: 0, Skipped: 999
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/779/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/779/Hazelcast-pr-builder-779.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientMapEventJournalBasicTest>AbstractEventJournalBasicTest.allowReadingWithFutureSeq:511->HazelcastTestSupport.assertOpenEventually:1222->HazelcastTestSupport.assertOpenEventually:1237 CountDownLatch failed to complete within 30 seconds, count left: 1
[INFO] 
[ERROR] Tests run: 24160, Failures: 1, Errors: 0, Skipped: 999
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

vbekiaris commented on hazelcast/src/test/java/com/hazelcast/test/starter/constructor/HazelcastExpiryPolicyConstructor.java:
```diff
@@ -0,0 +1,48 @@
+    Object createNew0(Object delegate) throws Exception {
+        Constructor<?> constructor = targetClass.getConstructor(Long.TYPE, Long.TYPE, Long.TYPE);
+
+        Object create = getFieldValueReflectively(delegate, "create");
```
```suggestion
        Duration create = getFieldValueReflectively(delegate, "create");
        Duration access = getFieldValueReflectively(delegate, "access");
        Duration update = getFieldValueReflectively(delegate, "update");

        Object[] args = new Object[]{
                create.getTimeUnit().toMillis(create.getDurationAmount()),
                access.getTimeUnit().toMillis(access.getDurationAmount()),
                update.getTimeUnit().toMillis(update.getDurationAmount()),
        };
```

`Duration` is a `javax.cache` API class so it will not be loaded by the `HazelcastAPIDelegatingClassLoader` and we can use it directly here. Apart from the `durationAmount` part we also need to take into account the `timeUnit` part to get an accurate clone. Minimal test:

```
        HazelcastExpiryPolicy policy = new HazelcastExpiryPolicy(3L, 5L, 4L, TimeUnit.SECONDS);
        HazelcastExpiryPolicyConstructor constructor = new HazelcastExpiryPolicyConstructor(HazelcastExpiryPolicy.class);
        HazelcastExpiryPolicy cloned = (HazelcastExpiryPolicy) constructor.createNew(policy);
        assertEquals(policy, cloned);
```

mmedenjak replied:
Strike two with using JCache classes eagerly where I was not supposed to (I remember one time where I was resolving some classes during node startup :)). Will update the code to what you suggested.

mmedenjak replied:
Misunderstood the first time but now it's clear. Thanks for the suggestion. Added a test for the constructor as well and applied changes to 3.12.z PR.

vbekiaris commented on hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/EventServiceImpl.java:
```diff
@@ -583,7 +583,8 @@ private void sendEvent(Address subscriber, EventEnvelope eventEnvelope, int orde
             Packet packet = new Packet(serializationService.toBytes(eventEnvelope), orderKey)
-                    .setPacketType(Packet.Type.EVENT);
+                    .setPacketType(Packet.Type.EVENT)
+                    .raiseFlags(Packet.FLAG_4_0);
```
minor: could move raising `FLAG_4_0` to `Packet` constructor as done in other PR

mmedenjak replied:
Sure, I was planning on doing it 👍

mmedenjak replied:
Done, updated in this and the 3.12.z PR. Now I only had to unset the flag when sending to 3.12 members and set the FLAG_3_12 flag, which is kinda ugly but the rest of the code using Packets is now unaffected.

vbekiaris commented on hazelcast/src/test/java/com/hazelcast/test/starter/constructor/AbstractConfigConstructor.java:
```diff
@@ -241,9 +253,306 @@ private static Object cloneSplitBrainProtectionFunctionImplementation(Object spl
     }
 
+    /**
+     * Clones the built-in QuorumFunction implementations.
```
minor: clones `EvictionConfig` actually

mmedenjak replied:
Fixed in both this and the 3.12.z PR.

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/internal/compatibility/cache/CompatibilityCacheDataSerializerHook.java:
```diff
@@ -0,0 +1,124 @@
+    public static final int SET_EXPIRY_POLICY = 67;
+    public static final int SET_EXPIRY_POLICY_BACKUP = 68;
+    public static final int EXPIRE_BATCH_BACKUP = 69;
+    public static final int CACHE_DESTROY_EVENT_CONTEXT = 70;
```
Yeah, I could not decide if I should remove ALL unused constants or leave them, in all of the factories and helper classes. I guess I left them in case I needed them in the future when I would discover I forgot to cover some edge case.

mmedenjak replied:
Done, removed all unused constants in a separate commit.

mmedenjak commented on hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/ByteArraySerializerAdapterTest.java:
```diff
@@ -61,7 +61,7 @@ public void testAdaptor() throws Exception {
 
         ByteArrayObjectDataOutput out = new ByteArrayObjectDataOutput(10, mockSerializationService, ByteOrder.BIG_ENDIAN);
-        ByteArrayObjectDataInput in = new ByteArrayObjectDataInput(out.buffer, mockSerializationService, ByteOrder.BIG_ENDIAN);
+        ByteArrayObjectDataInput in = new ByteArrayObjectDataInput(out.buffer, mockSerializationService, ByteOrder.BIG_ENDIAN, false);
```
Sure, I guess I preferred all code to be explicit about what they want but avoiding unnecessary changes works fine too :)

mmedenjak replied:
Done, added new constructors so that all of the existing code is unaffected.

mmedenjak commented on hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastStarterUtils.java:
```diff
@@ -202,4 +249,22 @@ public static void assertInstanceOfByClassName(String className, Object object)
+        } else if (HashMap.class.isAssignableFrom(type)) {
+            return new HashMap<Object, Object>();
+        } else if (Map.class.isAssignableFrom(type)) {
+            return new ConcurrentHashMap<Object, Object>();
```
I think I copied that one from the previous compatibility code, not sure what's the reason behind it but I guess it was just trying to be on the safe side - if we still haven't figured out which map the user is sending across classloaders, better make it a thread-safe map in case it's going to be used from different threads. Later I added the cases for specific implementations so this clause is probably going to be a lot less used.

mmedenjak commented on hazelcast/src/test/java/com/hazelcast/test/starter/ReflectionUtils.java:
```diff
@@ -176,6 +198,55 @@ public static void setFieldValueReflectively(Object arg, String fieldName, Objec
+        try {
+            method.invoke(methodObj, methodParam);
+        } catch (IllegalAccessException | InvocationTargetException | IllegalArgumentException e) {
+            debug("Could not update config via %s: %s", method.getName(), e.getMessage());
```
Will do!

mmedenjak replied:
Done, changed in both this and the 3.12.z PR.

mmedenjak commented:
@vbekiaris @blazember addressed all review comments (except the JCache one, I'll do that soon) and did the same changes on the 3.12.z PRs so it remains in-sync :)

Now, I'll attempt to rebase these commits so that the PRs are pointing towards the correct branches and I'll try to keep the same commits so you know what you reviewed and didn't review.

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/983/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/983/Hazelcast-pr-builder-983.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   PacketTest.setPacketType:64 expected same:<COMPATIBILITY_BIND_MESSAGE> was not:<BIND>
[INFO] 
[ERROR] Tests run: 24161, Failures: 1, Errors: 0, Skipped: 999
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/985/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/985/Hazelcast-pr-builder-985.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] COMPILATION ERROR : 
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/internal/memory/impl/BaseMemoryAccessorTest.java:[28,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:2.5.1:testCompile (default-testCompile) on project hazelcast: Compilation failure: Compilation failure: 
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/internal/memory/impl/BaseMemoryAccessorTest.java:[28,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/internal/memory/impl/BaseMemoryAccessorTest.java:[28,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/internal/memory/impl/BaseMemoryAccessorTest.java:[28,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/internal/memory/impl/BaseMemoryAccessorTest.java:[28,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/internal/memory/impl/BaseMemoryAccessorTest.java:[51,20] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/test/starter/constructor/HazelcastExpiryPolicyConstructor.java:[36,8] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/internal/memory/impl/BaseMemoryAccessorTest.java:[28,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/internal/memory/impl/BaseMemoryAccessorTest.java:[28,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/internal/memory/impl/BaseMemoryAccessorTest.java:[28,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/internal/memory/impl/BaseMemoryAccessorTest.java:[28,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/internal/memory/impl/BaseMemoryAccessorTest.java:[51,20] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/test/starter/constructor/HazelcastExpiryPolicyConstructor.java:[36,8] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/test/starter/constructor/HazelcastExpiryPolicyConstructor.java:[37,8] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/test/starter/constructor/HazelcastExpiryPolicyConstructor.java:[38,8] error: cannot find symbol
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/test/starter/constructor/HazelcastExpiryPolicyConstructor.java:[48,17] error: variable args is already defined in method createNew0(Object)
--------------------------
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/987/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/987/Hazelcast-pr-builder-987.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   DataSerializableConventionsTest.test_identifiedDataSerializables_areInstancesOfSameClass_whenConstructedFromFactory:265 Factory with ID 0 returned null for type with ID 44
[ERROR]   DataSerializableConventionsTest.test_identifiedDataSerializables_haveUniqueFactoryAndTypeId:191 Factory-Type ID pair {-31, 0} from class com.hazelcast.internal.compatibility.wan.CompatibilityWanReplicationEvent is already registered in another type.
[ERROR] Errors: 
[ERROR]   ProbabilisticSplitBrainProtectionFunctionTest.testSplitBrainProtectionAbsent_whenHeartbeatsLate:67->createSplitBrainProtectionFunctionProxy:120 ? ClassNotFound
[ERROR]   ProbabilisticSplitBrainProtectionFunctionTest.testSplitBrainProtectionPresent_whenHeartbeatsOnTime:79->createSplitBrainProtectionFunctionProxy:120 ? ClassNotFound
[ERROR]   RecentlyActiveSplitBrainProtectionFunctionTest.testSplitBrainProtectionAbsent_whenHeartbeatsReceivedBeforeToleratedWindow:70->createSplitBrainProtectionFunctionProxy:78 ? ClassNotFound
[INFO] 
[ERROR] Tests run: 3802, Failures: 2, Errors: 3, Skipped: 14
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1024/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1024/Hazelcast-pr-builder-1024.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   PartitionMigrationListenerTest.testMigrationStats_whenMigrationProcessRestarts:141->assertMigrationEventsConsistentWithResult:233 MigrationEvent{state=MigrationState{startTime=1590563316394, plannedMigrations=100, completedMigrations=1, totalElapsedTime=153ms}, partitionId=0, replicaIndex=0, source=Member [127.0.0.1]:5701 - 0ba4709a-ffa9-47ae-89a9-7a86e3999ca5 this, destination=Member [127.0.0.1]:5702 - cfdae9f8-7ccb-4207-87b5-cb2004e59ce4, success=false, elapsedTime=153ms}
[INFO] 
[ERROR] Tests run: 24162, Failures: 1, Errors: 0, Skipped: 999
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

mmedenjak commented:
All EE tests passed (http://jenkins.hazelcast.com/job/matko-Hazelcast-EE/45/) and all WAN compatibility tests passed (http://jenkins.hazelcast.com/job/matko-wan-migration-tool-tests/51/), merging. Thank you for the reviews!
